icu4c/source/i18n/number_rounding.cpp

```diff
@@ -36,27 +36,26 @@ void number::impl::parseIncrementOption(const StringSegment &segment,
     // Utilize DecimalQuantity/decNumber to parse this for us.
     DecimalQuantity dq;
     UErrorCode localStatus = U_ZERO_ERROR;
-    DecNum decnum;
-    decnum.setTo({buffer.data(), buffer.length()}, localStatus);
-    dq.setToDecNum(decnum, localStatus);
-    if (U_FAILURE(localStatus) || decnum.isSpecial()) {
+    dq.setToDecNumber({buffer.data(), buffer.length()}, localStatus);
+    if (U_FAILURE(localStatus) || dq.isNaN() || dq.isInfinite()) {
         // throw new SkeletonSyntaxException("Invalid rounding increment", segment, e);
         status = U_NUMBER_SKELETON_SYNTAX_ERROR;
         return;
     }
-    double increment = dq.toDouble();
-
-    // We also need to figure out how many digits. Do a brute force string operation.
-    int decimalOffset = 0;
-    while (decimalOffset < segment.length() && segment.charAt(decimalOffset) != '.') {
-        decimalOffset++;
+    bool sawDecimalPoint = false;
+    int32_t magnitude;
+    for (int32_t i=0; i<buffer.length(); i++) {
+        if (buffer[i] == '.') {
+            sawDecimalPoint = true;
+            magnitude = i - buffer.length() + 1;
+            dq.adjustMagnitude(-magnitude);
+            break;
+        }
     }
-    if (decimalOffset == segment.length()) {
-        outPrecision = Precision::increment(increment);
-    } else {
-        int32_t fractionLength = segment.length() - decimalOffset - 1;
-        outPrecision = Precision::increment(increment).withMinFraction(fractionLength);
+    if (!sawDecimalPoint) {
+        magnitude = dq.adjustToZeroScale();
```

Comment:
I am lost here. How is magnitude  stored in DecimalQuantity ? I saw you call dq.adjustMagnitude(-magnitude); which flip the magnitude while you pass it to dp, but then in here you took the return value of magnitude from dp w/o flipping it.

Reply:
I rearranged the code and added more comments to make it more clear what it is doing.